src/js/game/map_chunk.js

```diff
@@ -198,36 +198,35 @@ export class MapChunk {
             weights[enumSubShape.windmill] = 0;
         }
 
-        if (distanceToOriginInChunks < 10) {
-            // Initial chunk patches always have the same shape
-            const subShape = this.internalGenerateRandomSubShape(rng, weights);
-            subShapes = [subShape, subShape, subShape, subShape];
-        } else if (distanceToOriginInChunks < 15) {
-            // Later patches can also have mixed ones
-            const subShapeA = this.internalGenerateRandomSubShape(rng, weights);
-            const subShapeB = this.internalGenerateRandomSubShape(rng, weights);
-            subShapes = [subShapeA, subShapeA, subShapeB, subShapeB];
-        } else {
-            // Finally there is a mix of everything
-            subShapes = [
-                this.internalGenerateRandomSubShape(rng, weights),
-                this.internalGenerateRandomSubShape(rng, weights),
-                this.internalGenerateRandomSubShape(rng, weights),
-                this.internalGenerateRandomSubShape(rng, weights),
-            ];
-        }
-
         // Makes sure windmills never spawn as whole
-        let windmillCount = 0;
-        for (let i = 0; i < subShapes.length; ++i) {
-            if (subShapes[i] === enumSubShape.windmill) {
-                ++windmillCount;
+        let windmillCount;
+        do {
+            windmillCount = 0;
+            if (distanceToOriginInChunks < 10) {
+                // Initial chunk patches always have the same shape
+                weights[enumSubShape.windmill] = 0; // no whole windmills
```

Comment:
Why set this value to `0` again? The `if (distanceToOriginInChunks < 7)` above already changes it

Reply:
The distance is not increased. windmills and stars are still disabled below 7 distance like the original.

However, I added a line to disable windmills as well below 10 distance since only whole shapes can spawn and we never want whole windmill shapes. So taking out the chance of windmills avoids additional loops (since it would proc windmillCount > 2). I can of course just take out this line if you prefer.

Reply:
Oh, right, I forgot about that. Maybe just move it out since you don't have to repeatedly set it

(Not directly related to this issue, but what if 2 quads were just randomly chosen and replaced with something else? This changes the distributions, though)

Reply:
The distribution of shapes will be impacted yes. I did some calculations on my current solution:
distance < 10: Windmills are removed (-100%)
distance < 15: Windmill amount is greatly increased (+83-81%)
distance > 15: Windmill amount is slightly increased (+5-8%)

The reason for this drastic change between 10 and 15 is because in the previous algorithm, when a half shape would spawn with one half windmill shape it has a 50% chance of being overridden with the rectangle, depending on if the windmill generated left or right. I assume again unintentional behaviour.

I'm not sure if you are 100% set on keeping the current balance or not. I assume the changes at <10 and <15 basically cancel each other out. From personal experience, I wouldn't mind the 5% increase at further distances at all but I understand changing balance can have severe implications and are probably not decided by me. ><

I'll make a fix that keeps the old numbers for now, but I assume a change in the actual modifiers is required instead of this.

Reply:
I honestly don't care too much about the probabilities, as long as they are somewhat close to the original amounts (and twice as many is still relatively close)